File: 37/cpp/main.cc
#include <cassert>
#include <ctime>
#include <iostream>
#include <utility>
#include <vector>

#include "user_input.h"

namespace {

int generate_ramdom_index(int max) {
    static bool seed = false;
    if (!seed) {
        std::srand(std::time(nullptr));
        seed = true;
    }
    return std::rand() % max;
}

bool validate_number(const std::string& s, int& value) {
    try {
        // 数字であるかをチェック
        value = std::stoi(s);
        return true;
    } catch (std::invalid_argument e) {
        return false;
    } catch (std::out_of_range e) {
        return false;
    }
}

std::string generate_password(int num_special, int num_number, int num_min) {
    assert(num_special + num_number <= num_min);

    std::string alphabets =
        "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz";
    std::string numbers = "0123456789";
    std::string special_chars = "!\"#$%&'()*+,-./:;<=>?@[\\]^_`{|}~}";

    // パスワードに使う文字を集める
    std::vector<char> characters_selected;
    for (int i = 0; i < num_special; i++) {
        characters_selected.push_back(
            special_chars[generate_ramdom_index(special_chars.length())]);
    }
    for (int i = 0; i < num_number; i++) {
        characters_selected.push_back(
            numbers[generate_ramdom_index(numbers.length())]);
    }
    for (int i = 0; i < num_min - num_special - num_number; i++) {
        characters_selected.push_back(
            alphabets[generate_ramdom_index(alphabets.length())]);
    }

    // ランダムに並べ替える
    int remain = characters_selected.size();
    int iter_max = characters_selected.size() - 1;
    for (int i = 0; i < iter_max; i++) {
        int ridx = generate_ramdom_index(remain);
        // 入れ替える
        std::swap(characters_selected[i], characters_selected[i + ridx]);
        remain--;
    }
    return std::string(characters_selected.begin(), characters_selected.end());
}

}  // namespace

int main() {
    auto invalid_msg = "Sorry. That's not a valid input.";
    int num_min = read_user_input_with_validator<int>(
        "What's the minimum length? ", validate_number, invalid_msg);

    int num_special = read_user_input_with_validator<int>(
        "How many special characters? ", validate_number, invalid_msg);

    int num_number = read_user_input_with_validator<int>(
        "How many numbers? ", validate_number, invalid_msg);

    std::cout << generate_password(num_special, num_number, num_min)
              << std::endl;

    return 0;
}

File: 37/go/main.go
package main

import (
	"fmt"
	"math/rand"
	"strconv"
	"time"
)

const (
	alphabets    = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz"
	numbers      = "0123456789"
	specialChars = "!\"#$%&'()*+,-./:;<=>?@[\\]^_`{|}~}"
)

var r *rand.Rand

func init() {
	r = rand.New(rand.NewSource(time.Now().UnixNano()))
}

func generateRamdomNumber(max int) int {
	return r.Int() % max
}
func validateNumber(input string) (bool, int) {
	number, err := strconv.Atoi(input)
	if err != nil {
		return false, 0
	}
	return true, number
}

func generatePassword(numSpecial, numNumber, numMin int) string {
	if numSpecial+numNumber > numMin {
		panic("not enough length numMin")
	}

	// パスワードに使う文字を集める
	var charactersSelected []byte
	for i := 0; i < numSpecial; i++ {
		charactersSelected = append(charactersSelected, specialChars[generateRamdomNumber(len(specialChars))])
	}

	for i := 0; i < numNumber; i++ {
		charactersSelected = append(charactersSelected, numbers[generateRamdomNumber(len(numbers))])
	}

	for i := 0; i < numMin-numSpecial-numNumber; i++ {
		charactersSelected = append(charactersSelected, alphabets[generateRamdomNumber(len(alphabets))])
	}

	// ランダムに並べ替える
	remain := len(charactersSelected)
	iterMax := len(charactersSelected) - 1
	for i := 0; i < iterMax; i++ {
		rIdx := r.Int() % remain
		// 入れ替える
		charactersSelected[i], charactersSelected[i+rIdx] = charactersSelected[i+rIdx], charactersSelected[i]
		remain--
	}
	return string(charactersSelected)
}

func main() {

	invalidMsg := "Sorry. That's not a valid input."
	numMin := readUserInputWithValidator("What's the minimum length? ", validateNumber, invalidMsg)

	numSpecial := readUserInputWithValidator("How many special characters? ", validateNumber, invalidMsg)

	numNumber := readUserInputWithValidator("How many numbers? ", validateNumber, invalidMsg)

	password := generatePassword(numSpecial, numNumber, numMin)

	fmt.Printf("%s\n", password)
}
